raft/src/test/java/org/apache/kafka/raft/KafkaRaftLogTest.java

```diff
@@ -0,0 +1,1181 @@
+        MetadataLogConfig config = createMetadataLogConfig(
+                1024,
+                10 * 1000,
+                1024,
+                60 * 1000,
+                100,
+                DEFAULT_METADATA_LOG_CONFIG.internalMaxFetchSizeInBytes()
+        );
+        KafkaRaftLog log = buildMetadataLog(tempDir, mockTime, config);
+
+        for (int i = 0; i < 1000; i++) {
+            append(log, 1, 1);
+        }
+        log.updateHighWatermark(new LogOffsetMetadata(1001));
+
+        for (int offset : List.of(100, 200, 300, 400, 500, 600)) {
+            OffsetAndEpoch snapshotId = new OffsetAndEpoch(offset, 1);
+            createNewSnapshotUnchecked(log, snapshotId);
+        }
+
+        assertEquals(6, log.snapshotCount());
+        assertTrue(log.maybeClean());
+        assertEquals(1, log.snapshotCount(), "Expected only one snapshot after cleaning");
+        assertOptional(log.latestSnapshotId(), snapshotId ->
+            assertEquals(600, snapshotId.offset())
+        );
+        assertEquals(600, log.startOffset());
+    }
+
+    @Test
+    public void testSoftRetentionLimit() throws IOException {
+        // Set retention equal to the segment size and generate slightly more than one segment of logs
+        MetadataLogConfig config = createMetadataLogConfig(
+                10240,
+                10 * 1000,
+                10240,
+                60 * 1000,
+                100,
+                DEFAULT_METADATA_LOG_CONFIG.internalMaxFetchSizeInBytes()
+        );
+        KafkaRaftLog log = buildMetadataLog(tempDir, mockTime, config);
+
+        for (int i = 0; i < 2000; i++) {
+            append(log, 1, 1);
+        }
+        log.updateHighWatermark(new LogOffsetMetadata(2000));
+
+        // Then generate two snapshots
+        OffsetAndEpoch snapshotId1 = new OffsetAndEpoch(1000, 1);
+        RawSnapshotWriter snapshot = log.createNewSnapshotUnchecked(snapshotId1).get();
+        append(snapshot, 500);
+        snapshot.freeze();
+
+        // Then generate a snapshot
+        OffsetAndEpoch snapshotId2 = new OffsetAndEpoch(2000, 1);
+        snapshot = log.createNewSnapshotUnchecked(snapshotId2).get();
+        append(snapshot, 500);
+        snapshot.freeze();
+
+        // Cleaning should occur, but resulting size will not be under retention limit since we have to keep one snapshot
+        assertTrue(log.maybeClean());
+        assertEquals(1, log.snapshotCount(), "Expected one snapshot after cleaning");
+        assertOptional(log.latestSnapshotId(), snapshotId -> {
+            assertEquals(2000, snapshotId.offset(), "Unexpected offset for latest snapshot");
+            assertOptional(log.readSnapshot(snapshotId), reader ->
+                assertTrue(reader.sizeInBytes() + log.log().size() > config.retentionMaxBytes())
+            );
+        });
+    }
+
+    @Test
+    public void testSegmentsLessThanLatestSnapshot() throws IOException {
```

Comment:
Why did we skip `testSegmentMsConfigIsSetInMetadataLog()` in the old test ?

Reply:
This test was added after I opened the PR. I must have missed it while rebasing.